### DISCUSSION TITLE
Admin Console Init Failed

### DISCUSSION THREAD

**Discussion by mikepaxton** [iiab/iiab#3706] [General]
I've run into an issue where suddenly whenever I try to log into the Admin Console I get an error saying Init Failed.  
I found a couple previous issues/discussions listed in github that focus on this issue but none of them panned out.   
The first time (or two) I installed iiab it was installed on the SD card of the Pi3B+ and then moved to along with the rest of the rootfs to the external USB drive.  I noted once I moved iiab to the external drive I would get the init failed error.  So, on the next attempt I installed moved the rootfs from the SD to USB drive, then installed iiab.  Same result though, init failed.
The layout of this build:
Raspberry Pi3b+
External 5TB WD HDD
Booting off of external HD:  /sda1 - boot, /sda2 - rootfs, /sda3 - home, /sda4 - /library
So, my steps have been this.
1.  Install Bookwork PiOS - Lite on SD card.
2. Copy the rootfs from the SD card to /sda2
4. Change both fstab and cmdline.txt to use the external USB drive to from.
5. Once booting from USB is working, install iiab using 1-line-install
6. IIAB installs under /opt which is on /sda2 and library which is on /sda4.
7. Already had a bunch of zim files in /library/zim so iiab takes that over and uses it.
8. However, at this point when I try to login to the Admin Console I get the Init Failed error.

I've run ./iiab-cmdsrv3.py manually and have captured the output into file i'm including with the diagnostic file.  I also ran across this with iiab-summary, i'm not sure if it's saying iiab is looking for zims on /sda2 which would be wrong.  They are on /sda4.
iiab-summery

Filesystem Size Used Avail Use% Mounted on
/dev/sda2 29G 8.1G 20G 30% / ZIMs: 89 OER2Go: 1 Apps2B: 1

I think that covers most of what I've found, brain is starting to turn to mush.  I've posted the diag. to pastebin with ....initFailed_mdpaxton as the file. http://sprunge.us/hDf7Rk
<img width="1216" alt="Screenshot 2024-02-07 at 6 35 24 AM" src="https://github.com/iiab/iiab/assets/16995877/8a49feb7-69c8-4f77-973f-3fc4acf7dcfd">

**Comment by tim-moody** [comment]
Admin Console can be fooled by unexpected partition schemes in conjunction with external drives. I was expecting GET-EXTDEV-INFO to fail, but your screen shot and cmdsrv output do not show any errors. So probably the problem is on the front end. If you go into the browser javascript console (F12 on windows) do you see any errors (apart from missing icon file)?

Admin Console tries to decide where your primary install is and whether you have a secondary drive that might be a source of content for copying. You have an external drive that is primary, so it may get confused. Let me know what you find.

**Comment by mikepaxton** [comment]
Not seeing any errors but I do have 3 warnings.
There seems to be a number of them centered around the Zims which like I mentioned earlier are on a separate partition from rootfs.  Root being /sda2 and library is on /sda4.

<img width="892" alt="Screenshot 2024-02-08 at 10 17 07 AM" src="https://github.com/iiab/iiab/assets/16995877/fd48142e-367d-42a7-b0d2-d508c424ffad">

Edit:  Here is the full java console output.
http://sprunge.us/sv2J3z

**Comment by tim-moody** [comment]
We support /library on its own partition, but everything that belongs in library should be there. And /library must be mounted on /library, not USB0 or something.

where did your zims come from? did you run iiab-make-kiwix-lib after moving everything?

init is failing to load some catalogs and process them into javascript objects used for display.

in the js console you could try displaying the vars: kiwixCatalog, oer2goCatalog, langCodes, zimGroups, zimCategories, installedZimCatalog, externalZimCatalog

**Comment by tim-moody** [comment]
just to be clear zims are in /library/zims/content. (case sensitive)

If you are trying to get more content storage, I would boot from the sd card as before and mount sda4 on /library in fstab

**Comment by mikepaxton** [comment]
Hi Tim,

The zims are from a previous server, so when building this setup I had pre-copied them to the library partition.  All zims are located in /library/zims/content/.  I'm mounting library in fsab as "UUID=82599736-c47e-4141-8084-505b6edcdcdd       /library        ext4    defaults        0       2", so no usb.  

While running iiab-make-kiwix-lib i'm getting the following error:

<img width="1558" alt="Screenshot 2024-02-08 at 1 24 42 PM" src="https://github.com/iiab/iiab/assets/16995877/8eb9b6a4-96a8-421b-8988-5ded5f5541d1">

My primary purpose for using the external drive is to avoid the SD card which I've had really bad luck over the years with, as they tend to degrade over time.  So any permanent Pi system I build uses some form of external drive to boot from.

I should not, that despite iiab-make-kiwix-lib saying it finished building the Kiwix library.xml, I still get the init failed error.

**Comment by mikepaxton** [comment]
Just noticed something I should point out.  All of the zim files in /library/zims/content have root as both user and group.  Might this be the issue?

**Comment by mikepaxton** [comment]
Hi Tim,

Ok, so I moved all of the zims out of the content folder and ran iiab-make-kiwix-lib.  It rebuilt the file by removing all the zims from the file and completed successfully.  I was then able to log in to the admin-console.

So, I then moved one zim file "3dprinting.stackexchange.com_en_all_2023-07.zim" back into the content folder and ran iiab-make-kiwix-lib again.  It built the file, no issues and I was able to re-login to the Admin-Console, no issues.

I'm wondering if I have a corrupt zim file.  The odd thing about that is on my old server I can run kiwix-manage and rebuild the library with the exact same zim files.  

Thoughts?

**Comment by mikepaxton** [comment]
Hi Tim,

Solved it!  I started moving the zims back to the /content folder one by one and when I got to "gutenberg...zim", iiab-make-kiwix-lib failed to build.  I removed gutenberg and took a chance by moving all of the remainder back into /content and iiab-make-kiwix-lib worked just fine.  I'm able to log into Admin-Console, no issues.
It had to be the largest zim file of all of them, at almost 90Gb but I'll start the iiab download process on it right away.

Thanks so much for all your help in pointing me in the right direction.  Much appreciated.

Mike

**Comment by mikepaxton** [comment]
Just and fyi, I don't think it's a corrupt zim file.  I just noticed I had downloaded the Multi language version of it.  That might explain the error I was receiving with iiab-make-kiwix-lib:  
KeyError: 'eng,fra,fin,deu,nld,ita,spa,por,hun,zho,swe,ell,epo,lat,dan,tgl,cat,pol,nor,jpn,cym,afr,ces,isl,tel,heb,fur,bul,rus,ron,srp,fry,glg,enm,ilo,gla,gle,ceb,arp,slv,san,rmr,oji,oci,nap,nah,myn,mri,iku,ina,fas,est,bre,bod,ara'

**Comment by mikepaxton** [comment]
Sorry, one more question regarding iiab's Kiwix Server.  I've noticed the Home button seems to be missing from the build.

<img width="464" alt="Screenshot 2024-02-08 at 2 21 40 PM" src="https://github.com/iiab/iiab/assets/16995877/1991053f-9f78-465d-a6cf-1142d975e60c">

<img width="481" alt="Screenshot 2024-02-08 at 2 22 32 PM" src="https://github.com/iiab/iiab/assets/16995877/4ef8fc1f-beca-4517-9960-6d72b5ecde2c">


When you drill down a few pages into a zim, it's nice to be able to hit the Home button to get you back out the main Kiwix Server page.

**Comment by tim-moody** [comment]
Well the Home button is not there by design as I prefer IIAB navigation to Kiwix. I wish I could make the target of Home our top page.

**Reply by mikepaxton** [reply]
That's fair enough.

**Comment by tim-moody** [comment]
@mikepaxton I added a patch for the case where a zim has more than one language

https://github.com/iiab/iiab-admin-console/pull/562, now merged

could you possibly test it with your gutenberg...zim?

**Reply by mikepaxton** [reply]
Hi Tim,

Just got back in town and noticed your message.  
I just made your changes to the code in both "/opt/iiab/iiab-admin-console/roles/common/templates/adm_lib.py" and /usr/lib/python3/dist-packages/iiab/adm_lib.py then re-ran iiab-make-kiwix-lib.  It worked perfectly, made the library.xml without error.

**Reply by mikepaxton** [reply]
@tim-moody I didn't give you correct information regarding the multi-language kiwix fix.  The iiab-make-kiwix-lib worked fine, no error.  However, I totally forgot to try to log back into the admin-console.  In trying to do so I received the init failed again.  So I moved the multi back out of content and re-ran iiab-make-kiwix-lib, admin console works again.  So, the fix didn't solve the issue.

**Comment by tim-moody** [comment]
This is turning out to be more and more complicated. Kiwix used to just give language code as 'mul' for things that had sections in more than one language, but now someone has decided to provide a list of 30 or more for Phet and Gutenberg. I'm inclined not to support this as I don't think such content is very useful when the specific languages desired are available, are smaller, will have their own menu items, and will return search results that don't have cognates.

I can remove these from the catalog, but I'm not sure how to handle the case where someone downloads them manually.

**Reply by mikepaxton** [reply]
I totally agree with you.  In my case it was pure accident that I chose the multi version of Gutenberg.  I've switched to the english version.  So long as users have the option of downloading a specific language version of the zim file they want, I don't see a reason to support a multi language version.  That's my two cents on the subject.

**Reply by rgaudin** [reply]
@tim-moody FYI this is the [spec](https://wiki.openzim.org/wiki/Metadata)-recommended way and we are moving away from the `mul` trick to a list of language codes. Beware that `mul` will still be used in filenames and that we updated the spec to request that lang-codes list to be sorted by importance. So your patch above of using the first one is quite OK.

**Comment by tim-moody** [comment]
@rgaudin thanks for letting me know. So far the zims that have multiple languages are a combination of other zims that exist for separate languages. Is this likely to remain the case? So if I put the Gutenberg multi under English, will I have two there?

**Comment by tim-moody** [comment]
Is there any way we could have multi as a tag, like we have for _ftindex, etc.?

**Comment by rgaudin** [comment]
> So far the zims that have multiple languages are a combination of other zims that exist for separate languages. Is this likely to remain the case?

No. At the moment only phets and gutenberg are indeed using this but this will change.

If the content targets multiple languages, it gets (or will get) multiple codes in `Languages`. For instance, Stack Exchange ZIMs about _foreign_ languages like [russian](https://library.kiwix.org/viewer#russian.stackexchange.com_ru_all_2023-04/questions) are currently tagged as `rus` but it should probably be `eng,rus`.

> So if I put the Gutenberg multi under English, will I have two there?

yes.

> Is there any way we could have multi as a tag, like we have for _ftindex, etc.?

What would a multi be? If you mean `len(Language)>1` then you already have that information. 

I imagine you want to know _whether_ said ZIM is a multilingual collection of content that is also offered as per-language ZIM files in the Library. That's an interesting information that is _library-level_. I summon @kelson42 for this.

**Comment by tim-moody** [comment]
Out of curiosity would the order be eng,rus or rus,eng in your example?
I agree Language can be used to derive multi, though it is not as simple as you suggest because it is always a string > 1 char, not an array.

>I imagine you want to know whether said ZIM is a multilingual collection of content that is also offered as per-language ZIM files in the Library. That's an interesting information that is library-level. I summon @kelson42 for this.

I'm not asking for this, though I agree it would be useful.

**Reply by rgaudin** [reply]
> Out of curiosity would the order be eng,rus or rus,eng in your example?

It's because those decisions are not easy that we haven't switch them. As per the spec the "important" language for the ZIM should come first and it's vague on purpose. Personally I'd say eng,rus as I view it more oriented towards english speakers.


> I agree Language can be used to derive multi, though it is not as simple as you suggest because it is always a string > 1 char, not an array.

It's a list of comma separated values as per the spec. You can safely rely on this to turn it into an array.
